### DISCUSSION TITLE
Why has caf::string_view different constructors on different compilers?

### DISCUSSION THREAD

**Discussion by uentity** [actor-framework/actor-framework#1214] [Q&A]
The following
```C++
const auto s = "hello";
const auto sv = caf::string_view(s);
```
won't compile on Visual Studio, because `s` has type `const char* const`.

Why do we have conditional overloads instead of having both of them for every compiler?
```C++
#ifdef CAF_GCC
  constexpr string_view(const char* cstr) noexcept
      : data_(cstr),
        size_(strlen(cstr)) {
    // nop
  }
#else
  template <size_t N>
  constexpr string_view(const char (&cstr)[N]) noexcept
      : data_(cstr),
        size_(N - 1) {
    // nop
  }
#endif
```

**Comment by Neverlord** [comment]
> Why do we have conditional overloads instead of having both of them for every compiler?

Because in C++11 land (especially on older compilers), `constexpr` isn't that well supported. The first constructor for GCC only exists because older releases will refuse the array overload when using `constexpr`. We also can't have the first overload on all compilers because they will reject `strlen` in a `constexpr` setting.

If you have a C-string that you need to wrap into a `caf::string_view` then you'll have to explicitly call `caf::string_view{s, strlen(s)}`.

FWIW, `caf::string_view` is an intermediate solution and hopefully gets the axe sooner rather than later when dropping support for older compilers.
